### ISSUE TITLE
devcontainer configuration

### ISSUE THREAD

**Issue by maninalift** [SAFE-Stack/SAFE-template#344]
I notice that the Quick Start Guide doesn't give instructions for setting up using VS Code Remote Containers. I don't know whether this is intentional given that I couldn't get your `.devcontainer` configuration to work.

I created a `.devcontainer` config that is working for me. [maninalift/SAFE-devcontainer](https://github.com/maninalift/SAFE-devcontainer). If you like the look of it, I'm happy to create a pull request for the template library and the documentation.

I'm new not just to SAFE-stack but to F#. However, I'm looking forward to using it for a bunch of projects, thanks for your work.

**Comment by isaacabraham** [comment]
@maninalift what issues do you have with the bundled devcontainer? Maybe that's something we can fix quickly.

With regards to the template - we're doing a deep review of the template and considering how to move forward with it in a way that means the maintenance cost is not as high going forward. If the changes you're suggesting are just going to modify the devcontainer file itself I think we'd be delighted to accept a PR on that but if you think it's going to be more pervasive, I'd suggest we hold off until we move forward with the template for v2.

Hope that otherwise you're finding things to your liking!

**Comment by maninalift** [comment]
I don't remember specifically what errors I had. I can give it another go and let you know. 

Yes, it would be just for the devcontainer. However, I was about to post other questions relating to making the template simple and integrated, so I will hold off that. 

To be honest, I'm not sure the devcontainer belongs in the template. I guess it may be the simplest way for many who will have dotnet installed on the host system anyway but it seems wierd to bootstrap setting up a dotnet environment with dotnet.

**Comment by vrescobar** [comment]
Hi all,

I am also having problems with the dev container using the unmodified SAFE template. The error I am having is that the project "failed to load" without much explanation, ironically it worked well on the local environment (Mac OS). The log file was a long cryptic mess of file paths which began by:

```
Error:
MSBuild failed with exitCode 1 Working Directory: ....
```
First of all the dev container has sdk version 2 rather than 3, so `dotnet restore` could not work in any case because the server code depended on netcoreapp3.0. I am using right now the version suggested by @maninalift and I like from it that it uses a non root user which is always a good idea, no matter what.

After I tried the dev container suggested I still got an error when I tried to manually build from the container:
```
$ dotnet build src/Client/
Microsoft (R) Build Engine version 16.5.0+d4cbfca49 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 51.67 ms for /workspaces/XXXX/src/Client/Client.fsproj.
FSC : error FS0193: The module/namespace 'Encode' from compilation unit 'Thoth.Json' did not contain the val 'ValLinkagePartialKey(generateEncoder)' [/workspaces/secretar.io/src/Client/Client.fsproj]

Build FAILED.

FSC : error FS0193: The module/namespace 'Encode' from compilation unit 'Thoth.Json' did not contain the val 'ValLinkagePartialKey(generateEncoder)' [/workspaces/XXXX/src/Client/Client.fsproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:03.67
```
This error I could "solve" without completely understand the environment by executing on the container:
```
dotnet tool restore
dotnet paket add Thoth.Json.Net
dotnet paket restore
# dotnet build src/Client ### this one still fails with the same error
# However at this point the Loading problem disappears and intellisense starts working
dotnet build # works on both
```
For some time it seemed to work but as I served the files with the other Docker `dotnet fake build --target run` it ended up breaking the dependencies and now either seem to work.

TLDR: **for some reason both Dockerfiles are incompatible and I am still figuring out how to run the project while using a dev docker for a portable IntelliSense**

I think it makes sense to have a dev container as part as this template, it's about to have a headstart and even bulma templates are added to it. After all it is boiler plate no one likes to do when the goal is in fact to start a new project. I think it should however be properly documented, specially regarding permissions and also consistent with the Dockerfile used for deployment, including the user and permissions which are two very common errors.

If you make a PR I could review and test it as well.

**Comment by theimowski** [comment]
Thanks @vrescobar for reporting this - indeed we forgot about devcontainer when updated to net core 3.0. I'll see if I can make it work

**Comment by theimowski** [comment]
Updated .net core to 3.0 and name for the c# extension in dev container - can you try seeing if 1.22.4 works for you now?

**Comment by vrescobar** [comment]
I ended up solving my problem by updating the cotnainers to the 3.1 (see diff) and executing **only** the devcontainer from vscode, and all the *dotnet tools restore* and related operations from within the container using the own vscode terminal. Then to serve the contents also using only the vscode debug option provided (thank you for the preconf!)


I will try anyway the new version and tell how it goes.

```
diff --git a/.devcontainer/Dockerfile b/.devcontainer/Dockerfile
index 9b9a516..1c298bd 100644
--- a/.devcontainer/Dockerfile
+++ b/.devcontainer/Dockerfile
@@ -1,4 +1,4 @@
-FROM mcr.microsoft.com/dotnet/core/sdk:2.2
+FROM mcr.microsoft.com/dotnet/core/sdk:3.1
 
 # Add keys and sources lists
 RUN curl -sL https://deb.nodesource.com/setup_11.x | bash
diff --git a/Dockerfile b/Dockerfile
index 3dfff10..e6b0d01 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,4 +1,4 @@
-FROM mcr.microsoft.com/dotnet/core/aspnet:3.0-alpine
+FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-alpine
 COPY /deploy /
 WORKDIR /Server
 EXPOSE 8085
```

**Comment by vrescobar** [comment]
TL;DR: *Yes it runs perfect now.*
I am wondering why is it being used the 3.0 version rather than 3.1?

Because today I spent almost half day with issues regarding this docker and the "failed to load" error I mentioned previously I will document the exact steps I followed with the hope it will help someone else in the future. Feel free to reuse it for the README.

Steps:
- I created a new template with `dotnet new SAFE --layout fulma-admin --communication bridge --pattern streams --deploy docker` (this were the options I want for my project)
- Open vscode and accept to switch to the devcontainer and let it do the build. It completes successfully
- First issue things aren't restored yet (the readme does not mention to do it in the dev environment in case we use one)
<img width="264" alt="Screenshot 2020-04-02 at 23 24 22" src="https://user-images.githubusercontent.com/6759126/78301244-18548780-7539-11ea-8e9f-1f17f41e1788.png">
- Open integrated terminal (the shell is attached to the dev docker) and run `dotnet tool restore`
- At this point I can already run `dotnet paket restore` and all deps download properly.
- When it's done Visual Studio still won't properly work so I needed to reload the window (CMD + R or close and reopen).
- After a few seconds both projects are displayed as normal, IntelliSense works perfect, and if I have any compiler error it is showed at the "problems tab" of visual studio code. **Awesome!**
- Now I need to run the project in debug mode to see on whether everything works at the same time. I go to the "run" tab, select the option "Watch SAFE App" at the drop menu, and press the play. It takes a minute to download and compile everything but when it is done Server and Client code are built, executing and watching for changes at localhost:8080 so it is possible to live code and see the results immediately (try to change the html template and observe how things change in the html page)
<img width="355" alt="Screenshot 2020-04-02 at 23 41 33" src="https://user-images.githubusercontent.com/6759126/78302759-bb0e0580-753b-11ea-9016-837b112b4aa6.png">
- From this step right now I am not sure how to stop the running compilation without close vscode, that's something I will have to figure out later.

<img width="707" alt="Screenshot 2020-04-02 at 23 45 55" src="https://user-images.githubusercontent.com/6759126/78302970-20fa8d00-753c-11ea-89a1-fcc332facdfa.png">

Awesome job guys! Thanks to the SAFE team to take all that time configuring vscode, I think I never had such a good setup in other languages or IDEs.

From my side this issue can be closed already.
